### PR TITLE
Feature update/reset password

### PR DESCRIPTION
**Pull request by eoliveros**

### CODE REVIEW

File: paydb_endpoint.py [hunk 1]
```diff
@@ -1,4 +1,6 @@
 # pylint: disable=unbalanced-tuple-unpacking
+# pylint: disable=ungrouped-imports
```
Comment (djpnewton) [inline paydb_endpoint.py]: you should not need this, just group the imports

File: paydb_endpoint.py [hunk 2]
```diff
@@ -7,8 +9,9 @@
 
 from flask import Blueprint, request, jsonify, flash, redirect, render_template
 import flask_security
-from flask_security.utils import encrypt_password
+from flask_security.utils import encrypt_password, verify_password
 from flask_socketio import Namespace, emit, join_room, leave_room
+from flask_security.recoverable import send_reset_password_instructions
```
Comment (djpnewton) [inline paydb_endpoint.py:12]: move this up one line so the imports are grouped

File: paydb_endpoint.py [hunk 3]
```diff
@@ -243,6 +246,26 @@ def user_info():
         return jsonify(dict(email=user.email, balance=balance, photo=user.photo, photo_type=user.photo_type, roles=roles, permissions=perms))
     return jsonify(dict(email=user.email, balance=-1, photo=user.photo, photo_type=user.photo_type, roles=[], permissions=[]))
 
+@paydb.route('/user_reset_password', methods=['GET', 'POST'])
+def user_reset_password():
+    sig = request_get_signature()
+    content = request.get_json(force=True)
+    if content is None:
+        return bad_request(web_utils.INVALID_JSON)
+    params, err_response = get_json_params(content, ["api_key", "nonce", "email"])
```
Comment (djpnewton) [inline paydb_endpoint.py]: email is not used

File: paydb_endpoint.py [hunk 4]
```diff
@@ -288,6 +311,30 @@ def user_update_email_confirm(token=None):
...
+        return bad_request(web_utils.INVALID_JSON)
+    params, err_response = get_json_params(content, ["api_key", "nonce", "current_password", "new_password"])
+    if err_response:
+        return err_response
+    api_key, nonce, current_password, new_password = params
+    res, reason, api_key = check_auth(db.session, api_key, nonce, sig, request.data)
+    if not res:
+        return bad_request(reason)
+    user = api_key.user
+    verifying_password = verify_password(current_password, user.password)
```
Comment (djpnewton) [inline paydb_endpoint.py]: rename `verifying_password` to `verified_password` or `password_ok`

File: paydb_endpoint.py [hunk 5]
```diff
@@ -288,6 +311,30 @@ def user_update_email_confirm(token=None):
...
+    if not verifying_password:
+        return 'entered incorrect password'
```
Comment (djpnewton) [inline paydb_endpoint.py]: you need to return  `bad_request()`

**Review by djpnewton — CHANGES_REQUESTED** [review]

File: paydb_endpoint.py (hunk 1 above)
Comment (eoliveros) [reply paydb_endpoint.py]: done

File: paydb_endpoint.py (hunk 3 above)
Comment (eoliveros) [reply paydb_endpoint.py]: done

File: paydb_endpoint.py (hunk 4 above)
Comment (eoliveros) [reply paydb_endpoint.py]: done

File: paydb_endpoint.py (hunk 2 above)
Comment (eoliveros) [reply paydb_endpoint.py:12]: hmmm... this should already be there

File: paydb_endpoint.py (hunk 5 above)
Comment (eoliveros) [reply paydb_endpoint.py]: done

File: paydb_endpoint.py [hunk 6]
```diff
@@ -243,6 +245,26 @@ def user_info():
...
+@paydb.route('/user_reset_password', methods=['GET', 'POST'])
+def user_reset_password():
+    sig = request_get_signature()
+    content = request.get_json(force=True)
+    if content is None:
+        return bad_request(web_utils.INVALID_JSON)
+    params, err_response = get_json_params(content, ["api_key", "nonce"])
+    if err_response:
+        return err_response
+    ### email is set to None
```
Comment (djpnewton) [inline paydb_endpoint.py]: these next two lines are redundant

File: paydb_endpoint.py [hunk 7]
```diff
@@ -288,6 +310,30 @@ def user_update_email_confirm(token=None):
...
+    api_key, nonce, current_password, new_password = params
+    res, reason, api_key = check_auth(db.session, api_key, nonce, sig, request.data)
+    if not res:
+        return bad_request(reason)
+    user = api_key.user
+    verified_password = verify_password(current_password, user.password)
+    if not verified_password:
+        return bad_request('entered incorrect password')
+    ### set the new_password:
+    encrypted_new_password = encrypt_password(new_password)
```
Comment (djpnewton) [inline paydb_endpoint.py]: could compress these two lines into one

File: paydb_endpoint.py [hunk 8]
```diff
@@ -288,6 +310,30 @@ def user_update_email_confirm(token=None):
...
+    if err_response:
+        return err_response
+    api_key, nonce, current_password, new_password = params
+    res, reason, api_key = check_auth(db.session, api_key, nonce, sig, request.data)
+    if not res:
+        return bad_request(reason)
+    user = api_key.user
+    verified_password = verify_password(current_password, user.password)
+    if not verified_password:
+        return bad_request('entered incorrect password')
```
Comment (djpnewton) [inline paydb_endpoint.py]: make this string a constant in web_utils

File: test_api.py [hunk 9]
```diff
@@ -83,7 +94,6 @@ def construct_parser():
 
     parser_stash_save_check = subparsers.add_parser("stash_save_check", help="Check a user stash request")
     parser_stash_save_check.add_argument("token", metavar="TOKEN", type=str, help="The request token")
-
```
Comment (djpnewton) [inline test_api.py]: unneeded change

File: utils.py [hunk 10]
```diff
@@ -1,3 +1,4 @@
+# pylint: disable=missing-final-newline
```
Comment (djpnewton) [inline utils.py]: remove this, easier just to put a new line at the end of the file

**Review by djpnewton — CHANGES_REQUESTED** [review]

File: paydb_endpoint.py [hunk 11]
```diff
@@ -243,6 +246,24 @@ def user_info():
...
+    if err_response:
+        return err_response
+    api_key, nonce = params
+    res, reason, api_key = check_auth(db.session, api_key, nonce, sig, request.data)
+    if not res:
+        return bad_request(reason)
+    user = api_key.user
+    #### separate DB for the instructions(flask-security)
+    send_reset_password_instructions(user)
+    return 'Reset password instructions sent to '+str(user)+'.'
```
Comment (djpnewton) [inline paydb_endpoint.py]: just use 'reset password instructions sent'

File: paydb_endpoint.py [hunk 12]
```diff
@@ -81,6 +83,7 @@ def user_register():
     if not utils.is_email(email):
         return bad_request(web_utils.INVALID_EMAIL)
     if not password:
+# pylint: disable=missing-final-newline
```
Comment (djpnewton) [inline paydb_endpoint.py]: this should not be necessary

**Review by djpnewton — CHANGES_REQUESTED** [review]

File: paydb_endpoint.py (hunk 11 above)
Comment (eoliveros) [reply paydb_endpoint.py]: changed

File: paydb_endpoint.py (hunk 12 above)
Comment (eoliveros) [reply paydb_endpoint.py]: done

**Comment by djpnewton** [comment]
there are conflicts which need to be resolved

**Comment by eoliveros** [comment]
The merge didn't work correctly.... its not placed some of the changes i made. it looks correct on the new branch i merge with though. need to create a new PR for the new branch

**Comment by eoliveros** [comment]
Not sure how to link this up to the one below, its more up-to-date

https://github.com/zap-me/premio_stage/pull/13


closing this.

**Pull request closed without merging**